library/alloc/src/collections/btree/node.rs

```diff
@@ -1741,5 +1729,27 @@ unsafe fn slice_remove<T>(slice: &mut [MaybeUninit<T>], idx: usize) -> T {
     }
 }
 
+/// Transfers all elements from one slice to another, after which all elements
+/// in `source` should be considered uninitialized. If `dest` is longer than
+/// `source`, the extra elements in `dest` are left untouched.
+/// Somewhat resembles `dest.copy_from_slice()` but does not logically copy,
+/// does not require `T` to be `Copy`, does not require the slices to have the
+/// same lengths, and does not promise to panic if lengths are incompatible.
+///
+/// # Safety
+/// The `dest` slice is at least as long as the `source` slice.
+unsafe fn slice_move<T>(source: &mut [MaybeUninit<T>], dest: &mut [MaybeUninit<T>]) {
+    debug_assert!(dest.len() >= source.len());
+    unsafe {
+        ptr::copy_nonoverlapping(source.as_ptr(), dest.as_mut_ptr(), source.len());
+    }
+    if cfg!(debug_assertions) {
+        // Tell Miri that the copied slice became uninitialized, and zero it in debug builds.
```

Comment:
What do you mean by zero it in debug builds? This isn't zeroing anything...

Reply:
Probably not contractually, but in practice it zeroes, both in builds of alloc and [in this playground example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fec00574e79240334d10ee97546cab95).